pilot/pkg/config/kube/ior/route.go

```diff
@@ -531,19 +214,71 @@ func (r *route) createRoute(metadata config.Meta, gateway *networking.Gateway, o
 			TLS:            tlsConfig,
 			WildcardPolicy: wildcard,
 		},
-	}, metav1.CreateOptions{})
+	}
+}
+
+func (r *route) createRoute(
+	metadata config.Meta,
+	originalHost string,
+	tls *networking.ServerTLSSettings,
+	serviceNamespace, serviceName string,
+) (*v1.Route, error) {
+	IORLog.Debugf("creating route for hostname %s", originalHost)
+
+	nr, err := r.
+		routerClient.
+		Routes(serviceNamespace).
+		Create(context.TODO(), buildRoute(metadata, originalHost, tls, serviceNamespace, serviceName), metav1.CreateOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "error creating a route for the host %s from gateway: %s/%s",
+			originalHost,
+			metadata.Namespace,
+			metadata.Name)
+	}
+
+	IORLog.Infof("route %s/%s created for hostname %s from gateway %s/%s",
+		nr.ObjectMeta.Namespace, nr.ObjectMeta.Name,
+		nr.Spec.Host,
+		metadata.Namespace, metadata.Name)
+
+	return nr, nil
+}
+
+func (r *route) updateRoute(
+	metadata config.Meta,
+	originalHost string,
+	tls *networking.ServerTLSSettings,
+	serviceNamespace string, serviceName string,
+) (*v1.Route, error) {
+	IORLog.Debugf("updating route for hostname %s", originalHost)
+
+	nr, err := r.
+		routerClient.
+		Routes(serviceNamespace).
+		Update(context.TODO(), buildRoute(metadata, originalHost, tls, serviceNamespace, serviceName), metav1.UpdateOptions{})
 	if err != nil {
-		return nil, fmt.Errorf("error creating a route for the host %s (gateway: %s/%s): %s", originalHost, metadata.Namespace, metadata.Name, err)
+		return nil, errors.Wrapf(err, "error updating a route for the host %s from gateway: %s/%s",
+			originalHost,
+			metadata.Namespace,
+			metadata.Name)
 	}
 
-	IORLog.Infof("Created route %s/%s for hostname %s (gateway: %s/%s)",
+	IORLog.Infof("route %s/%s updated for hostname %s from gateway %s/%s",
 		nr.ObjectMeta.Namespace, nr.ObjectMeta.Name,
 		nr.Spec.Host,
 		metadata.Namespace, metadata.Name)
 
 	return nr, nil
 }
 
+func (r *route) findRoutes(metadata config.Meta) (*v1.RouteList, error) {
+	defaultLabelSet := getDefaultRouteLabelMap(metadata.Name, metadata.Namespace)
+
+	labels := labels.SelectorFromSet(defaultLabelSet)
+
+	return r.routerClient.Routes(metadata.Namespace).List(context.TODO(), metav1.ListOptions{LabelSelector: labels.String()})
```

Comment:
This is looking for Routes only in the namespace of the Gateway, but in my case, the Route is in `istio-system`.

Reply:
The `Route` is created at the namespace of where the `IngressGateway` service is. It is normally `istio-system`.

Reply:
Yes, but if I create the `Gateway` in `bookinfo`, while the `IngressGateway` service is in `istio-system`, the `Route` is in `istio-system` but this code tries to find it in `bookinfo`, since it takes the namespace of the `Gateway`, not of the `IngressGateway` service. 

Also be aware that the `IngressGateway` service might be in a different namespace (neither `bookinfo` nor `istio-system`).

Reply:
Some more info on this: I create a `Gateway` called `bookinfo-gateway` in `bookinfo`. The ingress gateway is in `istio-system`. If I then do:
```
kubectl -n bookinfo annotate gateway bookinfo-gateway foo=bar
```
the istiod log shows the following error:
```
error   ior     failed to process gateway bookinfo-gateway/bookinfo event update: 1 error occurred:
        * error creating a route for the host * from gateway: bookinfo/bookinfo-gateway: routes.route.openshift.io "bookinfo-bookinfo-gateway-684888c0ebb17f37" already exists
```

As I explained above, IOR thinks that the `Route` doesn't exist (because it's looking for it in `bookinfo`) and thus tries to create it (in `istio-system`). This fails, because the Route *does* exist there.

Reply:
It will look for all namespaces now. It uses the labels to narrow the matching results. The labels will include the name of the Gateway resource and the Namespace the Gateway is in.

Reply:
@luksa when I changed [line 260](https://github.com/maistra/istio/pull/673/files?diff=split&w=0#diff-d92eadb815953403f29e9eb8198b9b337121ecf2e8571c8f5d3bf0127f3a4353R260) to be 
```go
return r.routerClient.Routes(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{LabelSelector: labels.String()})
```

using `NamespaceAll` to select all routes and expect the label selector to narrow the result set. The service mesh integration failed. I will test it more.

Reply:
Why do you need to search in all namespaces? The `buildRoute()` function creates the Route in the `serviceNamespace` (the namespace of the ingress gateway pod). Can't you look for the Route only in that namespace?